File: desktop/src/lib/modelSearch.ts
// Pure logic + copy for the Model Discovery Hub (v1: search + link-out only).
// No JSX so filters, link-out URLs, and import-gating display are unit-testable.
// Source of truth: docs/design/MODEL_DISCOVERY_HUB.md.

export type ModelSource = "thingiverse" | "myminifactory" | "cults3d";

export interface ModelSearchResult {
  title: string;
  source: ModelSource;
  source_url: string;
  thumbnail_url?: string | null;
  license?: string | null;
  price_status?: "free" | "paid" | "unknown";
  file_formats?: string[];
  tags?: string[];
  attribution?: string | null;
  import_allowed: boolean;
  reason_import_not_allowed?: string;
}

export interface SearchResponse {
  results: ModelSearchResult[];
  providers_queried: string[];
  warnings: string[];
}

export interface SearchFilters {
  freeOnly?: boolean;
  commercialUse?: boolean;
  formats?: string[];          // e.g. ["STL","3MF"]
  multiColor?: boolean;
  sources?: ModelSource[];
}

export const SANCTIONED_SOURCES: ModelSource[] = ["thingiverse", "myminifactory", "cults3d"];

// Link-out-only providers: we never render normalized cards from them, only a
// button that opens their own search page.
export const LINK_OUT_PROVIDERS = [
  { id: "printables", label: "Printables" },
  { id: "thangs", label: "Thangs" },
  { id: "makerworld", label: "MakerWorld" },
] as const;

export const DISCLAIMER =
  "Models are provided by their source websites. Respect each model's license and creator terms.";

// Model Browser direction: the approved-site list a novice browses (link-out).
// Every entry has a home page and a search URL — Studio only ever opens these in
// the browser; it never scrapes, imports, or bypasses a site's terms.
export const BROWSE_PROVIDERS = [
  { id: "printables", label: "Printables", home: "https://www.printables.com" },
  { id: "thingiverse", label: "Thingiverse", home: "https://www.thingiverse.com" },
  { id: "myminifactory", label: "MyMiniFactory", home: "https://www.myminifactory.com" },
  { id: "cults3d", label: "Cults3D", home: "https://cults3d.com" },
  { id: "thangs", label: "Thangs", home: "https://thangs.com" },
  { id: "makerworld", label: "MakerWorld", home: "https://makerworld.com" },
] as const;

export function siteHomeUrl(source: string): string {
  return BROWSE_PROVIDERS.find((p) => p.id === source)?.home ?? "";
}

// Mirror of the Rust allowlist (desktop/src-tauri/src/main.rs). This is a
// frontend convenience check only — Rust is the real security boundary that
// blocks off-allowlist navigation in the in-app Model Browser.
export const APPROVED_MODEL_DOMAINS = [
  "printables.com", "thingiverse.com", "myminifactory.com",
  "cults3d.com", "thangs.com", "makerworld.com",
] as const;

export function isApprovedModelUrl(url: string): boolean {
  try {
    const u = new URL(url);
    if (u.protocol !== "https:") return false;
    const h = u.hostname.toLowerCase();
    return APPROVED_MODEL_DOMAINS.some((d) => h === d || h.endsWith(`.${d}`));
  } catch {
    return false;
  }
}

export function linkOutUrl(source: string, query: string): string {
  const q = encodeURIComponent(query || "");
  if (!q) return siteHomeUrl(source);
  switch (source) {
    case "printables": return `https://www.printables.com/search/models?q=${q}`;
    case "thingiverse": return `https://www.thingiverse.com/search?q=${q}&type=things`;
    case "myminifactory": return `https://www.myminifactory.com/search?query=${q}`;
    case "cults3d": return `https://cults3d.com/en/search?q=${q}`;
    case "thangs": return `https://thangs.com/search/${q}`;
    case "makerworld": return `https://makerworld.com/en/search/models?keyword=${q}`;
    default: return "";
  }
}

/** Plain-English label for why import is disabled (v1 disables all import). */
export function importReasonLabel(reason?: string): string {
  switch (reason) {
    case "paid": return "Paid model — open on the source site";
    case "no_download_api": return "No download API — open on the source site";
    case "license_unclear": return "License unclear — open on the source site";
    case "auth_required": return "Sign-in required on the source site";
    case "source_link_out_only":
    default: return "Import coming later — open on the source site";
  }
}

function isCommercialLicense(license?: string | null): boolean {
  if (!license) return false;
  const l = license.toLowerCase();
  // CC-BY (without NC) and explicit commercial terms allow commercial use.
  if (l.includes("nc") || l.includes("noncommercial") || l.includes("non-commercial")) return false;
  return l.includes("commercial") || l.includes("cc-by") || l.includes("cc by") || l.includes("standard");
}

/** Pure client-side filtering of normalized results. */
export function filterResults(results: ModelSearchResult[], filters: SearchFilters): ModelSearchResult[] {
  return results.filter((r) => {
    if (filters.freeOnly && r.price_status !== "free") return false;
    if (filters.commercialUse && !isCommercialLicense(r.license)) return false;
    if (filters.multiColor && !(r.tags ?? []).some((t) => /multi.?color|multicolour|amms?|mmu/i.test(t))) return false;
    if (filters.formats && filters.formats.length) {
      const fmts = (r.file_formats ?? []).map((f) => f.toUpperCase());
      if (!filters.formats.some((f) => fmts.includes(f.toUpperCase()))) return false;
    }
    if (filters.sources && filters.sources.length && !filters.sources.includes(r.source)) return false;
    return true;
  });
}

File: backend/snapstudio_core/model_search.py
"""Model Discovery Hub v1 — metadata search + link-out only.

Source of truth: docs/design/MODEL_DISCOVERY_HUB.md.

STRICT SCOPE for v1:
- Metadata search via SANCTIONED provider APIs only (Thingiverse, MyMiniFactory,
  Cults3D). No scraping. No third-party files. No paid downloads.
- Direct import is DISABLED in v1 (`import_allowed` is always False with a reason).
- Link-out-only providers (Printables, Thangs, MakerWorld) never produce
  normalized result cards — only a search URL to open on the source site.
- No API secrets here ship to the client; keys are read server-side from env. If a
  key is missing the provider is reported disabled — we never fake live results.
"""
from __future__ import annotations

import os
import urllib.parse

# Providers whose APIs we are permitted to query for normalized results.
SANCTIONED = ("thingiverse", "myminifactory", "cults3d")
# Providers we only ever link out to (their own search page), never normalize.
LINK_OUT_ONLY = ("printables", "thangs", "makerworld")

# Where each sanctioned provider's API key would come from (server-side only).
_KEY_ENV = {
    "thingiverse": "THINGIVERSE_APP_TOKEN",
    "myminifactory": "MYMINIFACTORY_API_KEY",
    "cults3d": "CULTS3D_API_KEY",
}


def assert_sanctioned(source: str) -> None:
    """No-scrape guard: refuse to normalize anything outside the sanctioned list."""
    if source not in SANCTIONED:
        raise ValueError(f"provider '{source}' is not a sanctioned search provider")


def _result(title, source, source_url, *, thumbnail_url=None, license=None,
            price_status="unknown", file_formats=None, tags=None, attribution=None) -> dict:
    allowed, reason = compute_import_allowed(source, price_status, license)
    return {
        "title": title,
        "source": source,
        "source_url": source_url,
        "thumbnail_url": thumbnail_url,
        "license": license,
        "price_status": price_status,
        "file_formats": file_formats or [],
        "tags": tags or [],
        "attribution": attribution,
        "import_allowed": allowed,
        "reason_import_not_allowed": reason,
    }


def compute_import_allowed(source: str, price_status, license):
    """v1 gating. import_allowed defaults False; returns (allowed, reason)."""
    if source == "cults3d":
        return False, "no_download_api"          # files stay on Cults by design
    if price_status == "paid":
        return False, "paid"
    if not license or str(license).strip().lower() in {"", "unknown", "unclear"}:
        return False, "license_unclear"
    # Otherwise eligible in principle, but direct import is not implemented in v1.
    return False, "source_link_out_only"


# ---- provider normalizers (raw provider JSON -> ModelSearchResult) ----

def normalize_thingiverse(raw: dict) -> dict:
    assert_sanctioned("thingiverse")
    lic = raw.get("license")
    creator = raw.get("creator")
    return _result(
        raw.get("name") or raw.get("title") or "(untitled)",
        "thingiverse",
        raw.get("public_url") or raw.get("url") or "",
        thumbnail_url=raw.get("thumbnail"),
        license=lic,
        price_status="free",                      # Thingiverse is free hosting
        file_formats=raw.get("formats") or ["STL"],
        tags=raw.get("tags") or [],
        attribution=creator.get("name") if isinstance(creator, dict) else creator,
    )


def normalize_myminifactory(raw: dict) -> dict:
    assert_sanctioned("myminifactory")
    price = "paid" if raw.get("is_paid") or (raw.get("price") or 0) else "free"
    images = raw.get("images")
    designer = raw.get("designer")
    return _result(
        raw.get("name") or "(untitled)",
        "myminifactory",
        raw.get("url") or "",
        thumbnail_url=(images[0].get("thumbnail") if images else raw.get("thumbnail")),
        license=raw.get("license"),
        price_status=price,
        file_formats=raw.get("formats") or [],
        tags=raw.get("tags") or [],
        attribution=designer.get("name") if isinstance(designer, dict) else designer,
    )


def normalize_cults3d(raw: dict) -> dict:
    assert_sanctioned("cults3d")
    price = "paid" if (raw.get("price") or {}).get("cents") else ("free" if raw.get("free") else "unknown")
    creator = raw.get("creator")
    return _result(
        raw.get("name") or "(untitled)",
        "cults3d",
        raw.get("url") or "",
        thumbnail_url=(raw.get("illustrationImageUrl") or raw.get("thumbnail")),
        license=raw.get("license"),
        price_status=price,
        file_formats=raw.get("formats") or [],
        tags=raw.get("tags") or [],
        attribution=creator.get("nick") if isinstance(creator, dict) else creator,
    )


def link_out_url(source: str, query: str) -> str:
    """Build the provider's own search-page URL (link-out only; no fetching)."""
    q = urllib.parse.quote(query or "")
    return {
        "printables": f"https://www.printables.com/search/models?q={q}",
        "thangs": f"https://thangs.com/search/{q}",
        "makerworld": f"https://makerworld.com/en/search/models?keyword={q}",
    }.get(source, "")


def provider_enabled(source: str) -> bool:
    """A sanctioned provider is enabled only if its server-side key is present."""
    env = _KEY_ENV.get(source)
    return bool(env and os.environ.get(env))


def search(query: str, filters: dict | None = None) -> dict:
    """v1 search. With no API keys configured, returns no results plus a clear
    per-provider 'disabled' warning — never faked/scraped data."""
    filters = filters or {}
    results: list[dict] = []
    providers_queried: list[str] = []
    warnings: list[str] = []

    sources = filters.get("sources") or list(SANCTIONED)
    for src in sources:
        if src not in SANCTIONED:
            warnings.append(f"{src}: link-out only — open it on the source site")
            continue
        if not provider_enabled(src):
            warnings.append(f"{src}: search disabled (no API key configured)")
            continue
        providers_queried.append(src)
        # Live querying is wired per-adapter when a key exists; v1 ships the
        # adapters + normalizers + gating. No network call is made without a key.

    return {"results": results, "providers_queried": providers_queried, "warnings": warnings}
